network.go

```diff
@@ -273,8 +277,134 @@ func getInterface(netHandle *netlink.Handle, link netlink.Link) *pb.Interface {
+func (s *sandbox) updateRoutes(netHandle *netlink.Handle, requestedRoutes *pb.Routes) (resultingRoutes *pb.Routes, err error) {
+
+	if netHandle == nil {
+		netHandle, err = netlink.NewHandle()
+		if err != nil {
+			return nil, err
+		}
+		defer netHandle.Delete()
+	}
+
+	//If we are returning an error, return the current routes on the system
+	defer func() {
+		if err != nil {
+			resultingRoutes, _ = getCurrentRoutes(netHandle)
+		}
+	}()
+
+	//
+	// First things first, let's blow away all the existing routes.  The updateRoutes function
+	// is designed to be declarative, so we will attempt to create state matching what is
+	// requested, and in the event that we fail to do so, will return the error and final state.
+	//
+
+	initRouteList, err := netHandle.RouteList(nil, netlink.FAMILY_ALL)
+	if err != nil {
+		return nil, err
+	}
+	for _, initRoute := range initRouteList {
+		// don't delete routes associated with lo:
+		link, _ := netHandle.LinkByIndex(initRoute.LinkIndex)
+		if link.Attrs().Name == "lo" || link.Attrs().Name == "::1" {
+			continue
+		}
+
+		err = netHandle.RouteDel(&initRoute)
+		if err != nil {
+			//If there was an error deleting some of the initial routes,
+			//return the error and the current routes on the system via
+			//the defer function
+			return
+		}
+	}
+
+	//
+	// Set each of the requested routes
+	//
+	// First make sure we set the interfaces initial routes, as otherwise we
+	// won't be able to access the gateway
+	for _, reqRoute := range requestedRoutes.Routes {
+		if reqRoute.Gateway == "" {
+			err = s.updateRoute(netHandle, reqRoute, true)
+			if err != nil {
+				agentLog.WithError(err).Error("update Route failed")
+				//If there was an error setting the route, return the error
+				//and the current routes on the system via the defer func
+				return
+			}
+
+		}
+	}
+	// Take a second pass and apply the routes which include a gateway
+	for _, reqRoute := range requestedRoutes.Routes {
+		if reqRoute.Gateway != "" {
+			err = s.updateRoute(netHandle, reqRoute, true)
+			if err != nil {
+				agentLog.WithError(err).Error("update Route failed")
+				//If there was an error setting the route, return the
+				//error and the current routes on the system via defer
+				return
+			}
+		}
+	}
+
+	return requestedRoutes, err
+}
+
+//getCurrentRoutes is a helper to gather existing routes in gRPC protocol format
+func getCurrentRoutes(netHandle *netlink.Handle) (*pb.Routes, error) {
+
+	if netHandle == nil {
+		netHandle, err := netlink.NewHandle()
+		if err != nil {
+			return nil, err
+		}
+		defer netHandle.Delete()
+	}
+
+	var routes pb.Routes
+
+	finalRouteList, err := netHandle.RouteList(nil, netlink.FAMILY_ALL)
+	if err != nil {
+		return &routes, err
+	}
+
+	for _, route := range finalRouteList {
+		var r pb.Route
+		if route.Dst == nil {
+			r.Dest = ""
+		} else {
+			r.Dest = route.Dst.String()
+		}
+
+		r.Gateway = route.Gw.String()
+		if r.Gateway == emptyRouteAddr {
+			r.Gateway = ""
+		}
```

Comment:
If I understand this correctly, this should be equal to 

```
if route.Gw != nil {
    r.Gateway = route.Gw.String()
}
```

And `emptyRouteAddr` can be removed from everywhere

Reply:
@WeiZhang555  No, netlink returns a string of "<nil">

Reply:
`route.Gw` is type `net.IP`, when `route.Gw == nil`, `route.Gw.String() == "<nil>"` according to my test.

According to this,  in your codes:
```
+		r.Gateway = route.Gw.String()
+		if r.Gateway == emptyRouteAddr {
 +			r.Gateway = ""
 +		}
```
is equal to 
```
+		r.Gateway = route.Gw.String()
+		if route.Gw == nil {
 +			r.Gateway = ""
 +		}
```

r.Gateway has a default value of "" so we only need to handle situation of `if route.Gw != nil`. A simplified version is 
```
if route.Gw != nil {
    r.Gateway = route.Gw.String()
}
```

Hope I'm making myself clear

Reply:
Thanks for spelling it out -- you're right -- let me make this change now...